src/AggregateFunctions/AggregateFunctionArgMinMax.h

```diff
@@ -29,35 +31,59 @@ struct AggregateFunctionArgMinMaxData
     {
         return ResultData::allocatesMemoryInArena() || ValueData::allocatesMemoryInArena();
     }
+
+    static String name()
+    {
+        return StringRef(ValueData_t::name()) == StringRef("min") ? "argMin" : "argMax";
+    }
 };
 
 /// Returns the first arg value found for the minimum/maximum value. Example: argMax(arg, value).
-template <typename Data>
-class AggregateFunctionArgMinMax final : public IAggregateFunctionDataHelper<Data, AggregateFunctionArgMinMax<Data>>
+template <typename Data, bool tuple_argument>
+class AggregateFunctionArgMinMax final : public IAggregateFunctionDataHelper<Data, AggregateFunctionArgMinMax<Data, tuple_argument>>
 {
 private:
     const DataTypePtr & type_res;
     const DataTypePtr & type_val;
 
 public:
     AggregateFunctionArgMinMax(const DataTypePtr & type_res_, const DataTypePtr & type_val_)
-        : IAggregateFunctionDataHelper<Data, AggregateFunctionArgMinMax<Data>>({type_res_, type_val_}, {}),
+        : IAggregateFunctionDataHelper<Data, AggregateFunctionArgMinMax<Data, tuple_argument>>({type_res_, type_val_}, {}),
         type_res(this->argument_types[0]), type_val(this->argument_types[1])
     {
         if (!type_val->isComparable())
             throw Exception("Illegal type " + type_val->getName() + " of second argument of aggregate function " + getName()
                 + " because the values of that data type are not comparable", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
     }
 
-    String getName() const override { return StringRef(Data::ValueData_t::name()) == StringRef("min") ? "argMin" : "argMax"; }
+    String getName() const override { return Data::name(); }
 
     DataTypePtr getReturnType() const override
     {
+        if constexpr (tuple_argument)
+        {
+            return std::make_shared<DataTypeTuple>(DataTypes{this->type_res, this->type_val});
+        }
+
         return type_res;
     }
 
-    void add(AggregateDataPtr place, const IColumn ** columns, size_t row_num, Arena * arena) const override
+    static const auto & getArgumentColumns(const IColumn**& columns)
```

Comment:
Do we really need reference to pointer here?

Reply:
Yes, it doesn't compile otherwise.

Reply:
To be honest, this construction looks a bit weird. Maybe you can make a better implementation? Or at least write a comment why do we need it. Also, add a whitespace before `**&`.

Reply:
I look weird but it makes sense, since it references stack variable which comes from previous function. I copied this construction from `sumMap` implementation.

Reply:
The problem is that the argument columns are either `IColumn **` for plain argument, or `TupleColumns` for tuple arguments. We can return the first one by value, but not the second one.